cmd/k8s-operator/connector.go

```diff
@@ -0,0 +1,262 @@
+		cn.Finalizers = append(cn.Finalizers, FinalizerName)
+		if err := a.Update(ctx, cn); err != nil {
+			err = fmt.Errorf("failed to add finalizer: %w", err)
+			logger.Errorf("error adding finalizer: %v", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// At the momement SubnetRouter is the only component configurable via
+	// ConnectorSpec and a ConnectorSpec without a SubnetRouter is invalid- but
+	// that will change in the future - so run the cleanup and provision
+	// conditionally.
+	if cn.Spec.SubnetRouter != nil && len(cn.Spec.SubnetRouter.Routes) > 0 {
+		var sb strings.Builder
+		sb.WriteString(string(cn.Spec.SubnetRouter.Routes[0]))
+		for _, r := range cn.Spec.SubnetRouter.Routes[1:] {
+			sb.WriteString(fmt.Sprintf(",%s", r))
+		}
+		cidrsS := sb.String()
+		logger.Debugf("ensuring a subnet router is deployed")
+		err := a.maybeProvisionSubnetRouter(ctx, logger, cn, cidrsS)
+		if err != nil {
+			msg := fmt.Sprintf(messageSubnetRouterCreationFailed, cidrsS, err)
+			cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{
+				Ready:   metav1.ConditionFalse,
+				Reason:  reasonSubnetRouterCreationFailed,
+				Message: msg,
+			}
+			a.recorder.Eventf(cn, corev1.EventTypeWarning, reasonSubnetRouterCreationFailed, msg)
+			return reconcile.Result{}, err
+		} else {
+			cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{
+				Routes:  cidrsS,
+				Ready:   metav1.ConditionTrue,
+				Reason:  reasonSubnetRouterCreated,
+				Message: fmt.Sprintf(messageSubnetRouterCreated, cidrsS),
+			}
+		}
+
+	} else {
+		logger.Debugf("ensuring a subnet router is cleaned up if it was ever created")
+		if done, err := a.maybeCleanupSubnetRouter(ctx, logger, cn); err != nil {
+			msg := fmt.Sprintf(messageSubnetRouterCleanupFailed, err)
+			cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{
+				Routes:  "",
+				Ready:   metav1.ConditionUnknown,
+				Reason:  reasonSubnetRouterCleanupFailed,
+				Message: msg,
+			}
+			a.recorder.Eventf(cn, corev1.EventTypeWarning, reasonSubnetRouterCleanupFailed, msg)
+			return reconcile.Result{}, err
+		} else if !done {
+			logger.Debugf("cleanup not done yet, will retry...")
+			cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{
+				Routes:  "",
+				Ready:   metav1.ConditionUnknown,
+				Reason:  reasonSubnetRouterCleanupInProgress,
+				Message: msgSubnetRouterCleanupInProgress,
+			}
+			return reconcile.Result{Requeue: true}, nil
+
+		} else {
+			cn.Status.SubnetRouter = &tsapi.SubnetRouterStatus{}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (a *ConnectorReconciler) maybeCleanupSubnetRouter(ctx context.Context, logger *zap.SugaredLogger, cn *tsapi.Connector) (bool, error) {
+	// At the momement SubnetRouter is the only component configurable via
+	// ConnectorSpec and a ConnectorSpec without a SubnetRouter is invalid- but
+	// that will change in the future - so run the cleanup and provision
+	// conditionally already now.
+	if cn.Spec.SubnetRouter == nil {
+		return true, nil
```

Comment:
I think this will result in always returning true from line 171, which means the cleanup won't ever happen but we will consider it did.

Reply:
I am going to set the subnet router def as required in the CRD spec, so it won't be possible to apply a Connector without a subnet router or routes defined. Then we can remove this check altogether.